### PR TITLE
Add `max_pool3d` for MPS

### LINKED ISSUE #100674: Add support for MaxPool3D on the MPS backend
### 🚀 The feature, motivation and pitch

Currently, Pooling operations are only supported on MPS for 1D and 2D inputs. While MPS doesn't have native support for 3d pooling operations, it does support 4d pooling operations (e.g. [`maxPooling4DWithSourceTensor()`](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/3750695-maxpooling4dwithsourcetensor?language=objc)). 3d tensors can be expanded to become 4d tensors, passed to 4d pooling operations, and then squeezed back to 3d tensors. This wouldn't affect the accuracy of the results when compared to direct 3d pooling operations. So, we can start working to add support for 3d pooling operations on the MPS backend.

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen @razarmehr @abhudev @mattiaspaul

### Alternatives

_No response_

### Additional context

This feature, along with the introduction of 3d convolutions in #99246 would open up lots of possible model architectures for 3D inputs. 
Full disclosure: I already have a working implementation of MaxPool3D support for MPS, and @mattiaspaul already has something in the works to address this as well.

### LINKED ISSUE #158723: max_pool3d_with_indices
### 🚀 The feature, motivation and pitch

When working with 3d imagery:
ERROR: The operator 'aten::max_pool3d_with_indices' is not currently implemented for the MPS device. If you want this op to be considered for addition please comment on https://github.com/pytorch/pytorch/issues/141287 and mention use-case, that resulted in missing op as well as commit hash e2d141dbde55c2a4370fac5165b0561b6af4798b. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.

### Alternatives

`PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS. But this is extremely slow and inefficient on my dataset. 

### Additional context

_No response_

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156467

Fixes #100674

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -1383,6 +1383,59 @@ def test_randperm(self, device="mps"):
...
+
+        for dilation in [(1, 1), (1, 1, 1, 1)]:
+            with self.assertRaisesRegex(RuntimeError, "dilation must be either a single int, or a tuple"):
+                torch.nn.functional.max_pool3d(torch.randn((1,) * 4, device=device), (1, 1, 1), dilation=dilation)
+
+        for stride in [(1, 1), (1, 1, 1, 1)]:
+            with self.assertRaisesRegex(RuntimeError, "stride must either be omitted, a single int, or a tuple"):
+                torch.nn.functional.max_pool3d(torch.randn((1,) * 4, device=device), (1, 1, 1), stride=stride)
+
+    _test_max_pool3d_cases = [
```
Comment (malfet) [inline test/test_mps.py]: Shouldn't this already be covered by opinfo tests?

File: aten/src/ATen/native/mps/operations/Pooling.mm [hunk 2]
```diff
@@ -240,6 +250,205 @@ static void pool2d_template(const Tensor& input,
...
+                  input_strides,
+                  output_sizes,
+                  output_strides,
+                  indices_sizes,
+                  indices_strides,
+                  work_pooling_dim_indices,
+                  kernel_size_fixed,
+                  stride_fixed,
+                  padding_fixed,
+                  dilation_fixed);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Pooling.mm]: CPU and Metal shader can share trivially constructable headers (see example with UpSample.h)
Please consider adding something like that for maxpool

File: aten/src/ATen/native/mps/operations/Pooling.mm [hunk 3]
```diff
@@ -240,6 +250,205 @@ static void pool2d_template(const Tensor& input,
...
+      at::empty({static_cast<int64_t>(arrayref.size())},
+                TensorOptions().device(c10::kCPU).dtype(at::kLong).memory_format(at::MemoryFormat::Contiguous));
+  std::memcpy(tensor.data_ptr<int64_t>(), arrayref.data(), arrayref.size() * sizeof(int64_t));
+  return tensor;
+}
+
+// NOTE: output is only valid as long as the tensor stays alive and its shape
+// doesn't change.
+static IntArrayRef tensor_to_intarrayref(const Tensor& tensor) {
+  TORCH_INTERNAL_ASSERT(tensor.dim() == 1);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Pooling.mm:265]: Don't you want assert that tensor is CPU tensor? (otherway one can not deref its data_ptr)

File: aten/src/ATen/native/mps/operations/Pooling.mm [hunk 4]
```diff
@@ -240,6 +250,205 @@ static void pool2d_template(const Tensor& input,
...
+
+  IntArrayRef output_size = tensor_to_intarrayref(t_output_size);
+  output.resize_(output_size);
+  indices.resize_(output_size);
+
+  auto iter = TensorIteratorConfig().add_output(output).resize_outputs(false).check_all_same_dtype(false).build();
+
+  id<MTLDevice> device = MPSDevice::getInstance()->device();
+  MPSStream* mpsStream = getCurrentMPSStream();
+  const uint32_t numThreads = iter.numel();
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Pooling.mm]: This silently downcasts it to uint32 which can lead to silent correctness problem later. `mtl_dispatchThreads` accepts uint64 and errors out if it's greater than 2**32 elements
```suggestion
  const auto numThreads = iter.numel();
```

File: aten/src/ATen/native/mps/operations/Pooling.mm [hunk 5]
```diff
@@ -240,6 +250,205 @@ static void pool2d_template(const Tensor& input,
...
+  TORCH_INTERNAL_ASSERT(numThreads == output.numel());
+  const int64_t dims = input.dim();
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Pooling.mm]: Why int64? dim should be pretty small (Metal devices currently do not natively support int64, so all computations over int64 are 2x slower than over in32)

File: test/test_mps.py (hunk 1 above)
Comment (kurtamohler) [reply test/test_mps.py]: Yes for sure, I'll remove this

File: aten/src/ATen/native/mps/operations/Pooling.mm (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/operations/Pooling.mm]: That is better, I will do that

File: aten/src/ATen/native/mps/operations/Pooling.mm (hunk 5 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/operations/Pooling.mm]: Got it, I'll change to int32_t

File: aten/src/ATen/native/mps/kernels/Pooling.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+// N is the maximum allowed number of dimensions in the input and outputs. The
+// maximum allowed pooling dimensions is N-2, because the input may have up to 2
+// leading dimensions that are not pooled. To support up to 3-D pooling, N=5 is
+// the default.
+template <unsigned N = 5>
+struct PoolingParams {
+  int32_t dims;
+  int32_t pooling_dims;
+  _ARRAY_NS::array<int64_t, N> input_sizes;
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/kernels/Pooling.h:19]: One small thing to note about providing these parameters in a struct is that the buffers in the struct need to be a fixed size at compile time (if I understand correctly), whereas before I made this change, the sizes could be determined at runtime. So this puts a limit on how many dimensions we can support while reusing the same code. But at the moment, pytorch doesn't support anything higher than 3-D pooling with 2 leading dims and I'm guessing there aren't any plans to support any higher dimensionality, so it's probably not really an issue. Of course if an arbitrary number of dims are needed in the future, we can just switch it back to specifying these params directly to `mtl_setArgs` instead of in a struct

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/mps/kernels/Pooling.h (hunk 6 above)
Comment (malfet) [reply aten/src/ATen/native/mps/kernels/Pooling.h:19]: Yes, this is correct, but even for total number of dims there is an upper limit on total number of tensor dimensions, which I think is currently 16

Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/Pooling.h:19]: Oh I didn't realize that, good to know
